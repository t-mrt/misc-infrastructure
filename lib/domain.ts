import * as cdk from "@aws-cdk/core"
import * as r from "@aws-cdk/aws-route53"
import * as t from "@aws-cdk/aws-route53-targets"
import * as c from "@aws-cdk/aws-cloudfront"

// TODO: divide stack by hosted zone

export class DomainStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props)
        const zone = new r.PublicHostedZone(this, "yux3", {
            zoneName: "yux3.net",
        })
        new r.CnameRecord(this, "blog", {
            zone: zone,
            domainName: "hatenablog.com.",
            recordName: "blog",
        })
        const distribution = c.Distribution.fromDistributionAttributes(this, "ImportedDist", {
            domainName: "d25ymprel65g4i.cloudfront.net",
            distributionId: "E1AB8RFEZ1H7F0",
        })
        new r.ARecord(this, "img", {
            zone: zone,
            target: r.RecordTarget.fromAlias(new t.CloudFrontTarget(distribution)),
            recordName: "img",
        })
    }
}
