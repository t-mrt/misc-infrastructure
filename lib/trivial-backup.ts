import * as cdk from "@aws-cdk/core"
import * as iam from "@aws-cdk/aws-iam"
import * as s3 from "@aws-cdk/aws-s3"

export class TrivialBackupStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props)
        const user = new iam.User(this, "User", {
            userName: "trivial_backup",
        })

        const bucket = new s3.Bucket(this, "Bucket", {
            versioned: true,
            bucketName: "trivial-backup-02",
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
        })

        bucket.addToResourcePolicy(
            new iam.PolicyStatement({
                actions: ["s3:PutObject"],
                resources: [bucket.arnForObjects("*")],
                principals: [user],
            })
        )
    }
}
