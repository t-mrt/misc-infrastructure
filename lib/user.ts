import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam'

export class UserStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    const user = new iam.User(this, 'GeneralPurposeUser', {
      userName: "t_"
    });

    const role = new iam.Role(this, 'PowerUserRole', {
      assumedBy: new iam.ArnPrincipal(user.userArn).withConditions({
        Bool: {
          "aws:MultiFactorAuthPresent": "true"
        },
      }),
      roleName: "admin_role",
    });

    user.addToPolicy(new iam.PolicyStatement({
      resources: [role.roleArn],
      actions: ['sts:AssumeRole'],
    }))

    role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"))
  }
}
