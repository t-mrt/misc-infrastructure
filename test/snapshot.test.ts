import * as cdk from "@aws-cdk/core"
import { SynthUtils } from "@aws-cdk/assert"
import * as domain from "../lib/domain"
import * as user from "../lib/user"
import * as tb from "../lib/trivial-backup"

describe("snapshot", () => {
    test("domain", () => {
        const app = new cdk.App()
        const stack = new domain.DomainStack(app, "MyTestStack")
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
    })
    test("user", () => {
        const app = new cdk.App()
        const stack = new user.UserStack(app, "MyTestStack")
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
    })
    test("trivial-backup", () => {
        const app = new cdk.App()
        const stack = new tb.TrivialBackupStack(app, "MyTestStack")
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
    })
})
