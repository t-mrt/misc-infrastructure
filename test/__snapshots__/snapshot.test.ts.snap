// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot domain 1`] = `
Object {
  "Mappings": Object {
    "AWSCloudFrontPartitionHostedZoneIdMap": Object {
      "aws": Object {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": Object {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Resources": Object {
    "blogF3DAE94C": Object {
      "Properties": Object {
        "HostedZoneId": Object {
          "Ref": "yux326B34FD5",
        },
        "Name": "blog.yux3.net.",
        "ResourceRecords": Array [
          "hatenablog.com.",
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "img31CD142E": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": "d25ymprel65g4i.cloudfront.net",
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": Object {
          "Ref": "yux326B34FD5",
        },
        "Name": "img.yux3.net.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "yux326B34FD5": Object {
      "Properties": Object {
        "Name": "yux3.net.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
  },
}
`;

exports[`snapshot trivial-backup 1`] = `
Object {
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "trivial-backup-02",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "User00B015A1",
                    "Arn",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "User00B015A1": Object {
      "Properties": Object {
        "UserName": "trivial_backup",
      },
      "Type": "AWS::IAM::User",
    },
  },
}
`;

exports[`snapshot user 1`] = `
Object {
  "Resources": Object {
    "GeneralPurposeUser8600AB08": Object {
      "Properties": Object {
        "UserName": "t_",
      },
      "Type": "AWS::IAM::User",
    },
    "GeneralPurposeUserDefaultPolicy8C89C9A0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PowerUserRole25AA3B40",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GeneralPurposeUserDefaultPolicy8C89C9A0",
        "Users": Array [
          Object {
            "Ref": "GeneralPurposeUser8600AB08",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PowerUserRole25AA3B40": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Condition": Object {
                "Bool": Object {
                  "aws:MultiFactorAuthPresent": "true",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "GeneralPurposeUser8600AB08",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
        "RoleName": "admin_role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
